### PR TITLE
Support non-backtracking regex

### PR DESCRIPTION
**Pull request by dbalikhin**
Fixes: #510 

Opening a draft request.
Testing + tests + sanity checks remaining.

Summary:
New EnableNonBacktrackingRegex property that comes from CLI input. False by default.
During rule conversion to OAT, it will check for this flag and add `nb` modifier unless `b` is already specified in the rule.
Thus, a local `b` rule modifier will overwrite global non-backtracking setting (if we want to use NonBacktracking but some rules require backtracking).
RuleVerifier will try with NB and if it fails, try with Backtracking.

### CODE REVIEW

File: AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs
Comment (gfs) [inline AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs:89]: Im not sure this is the right place to add this. The validation delegate isn't guaranteed to run. Should likely be in the RegexModiferToRegexOptipns method instead.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (gfs) [inline AppInspector.RulesEngine/RulesVerifier.cs:155]: Same as above. This should be part of the conversion method.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (gfs) [inline AppInspector.RulesEngine/RulesVerifier.cs:170]: Should use regexOpts from before with

`regexOpts &= ~RegexOptions.NonBackTracking` right?

**Comment by gfs** [comment]
Will review more thoroughly this afternoon.

**Review by gfs — CHANGES_REQUESTED** [review]
Will do a more thorough review this afternoon. But a few things stood out to me here.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/RulesVerifier.cs:170]: regexOpts won't be available here due to the declaration scope. We will need to move it above `try` block, but we might have some exceptions from RegexModifierToRegexOptions method. So we will need to put it to another try.
Not sure if we are catching exceptions "higher".

File: AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs:89]: Probably, this is a new method to me, I was asking the same question myself.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/RulesVerifier.cs:155]: sure, additional optional parameter for RegexModifierToRegexOptions should work fine.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (gfs) [reply AppInspector.RulesEngine/RulesVerifier.cs:170]: I don't think `RegexModifierToRegexOptions` has any cases where it will throw, so it should be safe to move out of the try.

File: AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs
Comment (gfs) [reply AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs:89]: Yes, I just added this yesterday so understandable you may have missed it in the merge. I was hoping to make it a little easier to expand the modifiers available with the new method.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (gfs) [inline AppInspector.RulesEngine/RulesVerifier.cs:168]: Can't add this message to the errors. Any messages in the errors indicate validation failure. The Validation object doesn't currently have a field for informative messages like warnings, Errors are always assumed to be a validation failure.

Comment (gfs) [reply AppInspector.RulesEngine/RulesVerifier.cs:168]: Actually this might not be true. Double checking the `Verified` field of the `RuleStatus` object it only checks the `OatIssues` field not the errors field.

Comment (gfs) [reply AppInspector.RulesEngine/RulesVerifier.cs:168]: https://github.com/microsoft/ApplicationInspector/blob/c187a93438406c47be469285d731fca950d6d8df/AppInspector.RulesEngine/RuleStatus.cs#L11

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/RulesVerifier.cs:170]: Haha, you are right. For some reason I read the code below the actual method in the Utils.cs file. Will change!

File: AppInspector/Commands/PackRulesCommand.cs
Comment (dbalikhin) [reply AppInspector/Commands/PackRulesCommand.cs:24]: We do Rules Verification as a part of PackRules.  Rules Verification includes NB.
But I think you are right, and it should not be packed.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/RulesVerifier.cs:168]: Just log then.

File: AppInspector.RulesEngine/Ruleset.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/Ruleset.cs:20]: I'm going to make it public for now if you don't mind. You can refactor later, but we won't break an API contract.

File: AppInspector.Common/Utils.cs
Comment (gfs) [inline AppInspector.Common/Utils.cs:25]: I thinl this switch list is missing two cases now that should be handled. `b` in the case that non backtrack is set and `nb` in either case. The `b` can be set with the `&= ~` technique and the nb with `|=`. Specifically for cases where it's set individually in the rule rather than by parameter.

File: AppInspector.RulesEngine/AbstractRuleSet.cs
Comment (gfs) [inline AppInspector.RulesEngine/AbstractRuleSet.cs:84]: This is an interesting edge case we probably want to some how catch in the validator. B and nb are exclusive options.

**Comment by gfs** [comment]
Let me know when you're ready for me to do another review.

File: AppInspector.RulesEngine/AbstractRuleSet.cs
Comment (gfs) [reply AppInspector.RulesEngine/AbstractRuleSet.cs:84]: I also wonder, if `nb` is being added here, do we also need the parameter to `RegexModifierToRegexOptions` or is this nb enough to flow through?

File: AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs
Comment (gfs) [inline AppInspector.RulesEngine/OatExtensions/OatRegexWithIndexOperation.cs:86]: I think this is not necessary, since `nb` is added to the arguments when the pattern is added to the abstract rule set - as long as the `nb` and `b` cases are added to `RegexModifierToRegexOptions`

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (gfs) [inline AppInspector.RulesEngine/RulesVerifier.cs:161]:
```suggestion
                    regexOpts &= ~RegexOptions.NonBacktracking;
```

I think its a bit clearer to just not the specific bit rather than xoring.

**Review by gfs — COMMENTED** [review]
Overall looking pretty good I think. A few more comments. Before merging will need a few tests still.

File: AppInspector.RulesEngine/AbstractRuleSet.cs
Comment (gfs) [reply AppInspector.RulesEngine/AbstractRuleSet.cs:84]: Another comment on this line. `b` and `nb` should be checked case insensitively to match implementation in `RegexModifierToRegexOptions`.

Perhaps

`pattern.Modifiers.Any(modifier => modifier.Equals("b", StringComparison.CurrentCultureIgnoreCase)`

**Comment by gfs** [comment]
@dbalikhin Thanks for your work on this so far! I think its getting pretty close. Just let me know if you'd like me to help make the changes I proposed above, or if you have any concerns about them.

**Comment by dbalikhin** [comment]
hey @gfs, I will try to complete it on this weekend. If for some reason it is not done (well, family stuff), feel free to make changes.

File: AppInspector.Common/Utils.cs
Comment (dbalikhin) [reply AppInspector.Common/Utils.cs:22]: Changed to `&= ~`

File: AppInspector.RulesEngine/AbstractRuleSet.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/AbstractRuleSet.cs]: I rewrote this piece. I'm using string comparison InvariantCultureIgnoreCase similar to all cases in this file. Just keeping the same style.

File: AppInspector.RulesEngine/RulesVerifier.cs
Comment (dbalikhin) [reply AppInspector.RulesEngine/RulesVerifier.cs:161]: Yeah, changed the logic, a bit of edge case, but your option is better.

**Comment by dbalikhin** [comment]
I added some validation logic to AbstractRuleSet. It should handle mutually exclusive `b` and `nb` now.
I will review the tests project and create some tests.

**Comment by dbalikhin** [comment]
Hm, it does look like we don't have rule validation if it comes from a default ruleset.
Should we modify StringToRegex to catch this edge case or ruleverify command needs to be used every time before changing a default ruleset?
```csharp 
   private Regex? StringToRegex(string built, RegexOptions regexOptions)
    {
        if (!RegexCache.ContainsKey((built, regexOptions)))
        {
            try
            {
                RegexCache.TryAdd((built, regexOptions), new Regex(built, regexOptions));
            }
            catch (NotSupportedException)
            {
                _logger.LogWarning("Provided regex {Regex} was not supported and could not be used", built);
#if NET7_0_OR_GREATER
                regexOptions &= ~RegexOptions.NonBacktracking;
                RegexCache.TryAdd((built, regexOptions), new Regex(built, regexOptions));
#endif
```

Trying to write some test cases and got exceptions in StringToRegex.

**Comment by gfs** [comment]
When changing the default rule set I either run the verify command against the default rules or use the unit test that validates default rules (using the verify command under the hood).

Since we have the unit test in CI it seemed redundant to me to have analyze also validate the rules - as they are embedded in the binary so should not have changed from the build step to the publication.

**Comment by gfs** [comment]
When changing the default rule set I either run the verify command against the default rules or use the unit test that validates default rules (using the verify command under the hood).

Since we have the unit test in CI it seemed redundant to me to have analyze also validate the rules - as they are embedded in the binary so should not have changed from the build step to the publication. Verify does definitely have to be run every time there's a change to the default rules. 

As for the second question, I'm not totally sure what sequence of events is occurring you're having trouble with, what is the exception thrown?

**Comment by gfs** [comment]
I reread the above, and think I understand what you mean now. I don't think the absence of rule verification of the default rules on each run should be an issue, since they are verified in the build step via CI and are embedded in the binary.

**Comment by gfs** [comment]
@dbalikhin I added a slight modification to your proposed fix for `StringToRegex` and fixed up the multiple places where regexes were being created. I can revisit this next week to add any needed tests, but i think overall looks good and should be able to merge next week. Thanks again!

**Review by gfs — COMMENTED** [review]
LGTM, pending adding a few more test cases, I can address next week and aim to merge next week as well.

Some tests that would be good to add:
Test the Utils.StringToRegex method directly
Test the behavior of a ruleset with a rule not compatible with non backtracking with nonbacktracking set
Test that the rule verifier warns about having both `b` and `nb`, or if `nb` is set and the rule is not `nb` compatible.

File: AppInspector.Common/Utils.cs
Comment (gfs) [inline AppInspector.Common/Utils.cs:36]: Possibly will want to add an optional logger parameter to this to get information about what exceptions were found for the rule verifier.

File: AppInspector/Commands/VerifyRulesCommand.cs
Comment (gfs) [inline AppInspector/Commands/VerifyRulesCommand.cs:114]: Should add a specific check in here for incompatible `nb` and `b` both specified - and cases with Regex with `nb` explicitly set that cannot be constructed as `nb` (but not those which cannot be constructed with `nb` but only because the `EnableNonBacktrackingRegex` option was set.

**Comment by dbalikhin** [comment]
I added some tests, feel free to modify anything. 

> "the rule verifier warns about having both b and nb, or if nb is set and the rule is not nb compatible."
Probably rule conversion will handle these edge cases. I think it works just fine with the extra logic around RegexCache that you added. Check my tests, they aren't for the verify command, but a generated ruleset should work after.

NB enabled with b and nb: b overwrites nb
NB disabled with b and nb: nb overwrites b
NB is enabled/set, but the rule is not compatible, nb removed from this regex.

**Review by gfs — APPROVED** [review]
LGTM.

**Comment by gfs** [comment]
Awesome contribution @dbalikhin! Definitely appreciate your work here!  I plan to merge this this afternoon.

**Pull request merged**
